File: AZ-204_Implement_API_Management/code/resource_mgmnt.py
from azure.identity import DefaultAzureCredential
from azure.mgmt.apimanagement import ApiManagementClient

# Set your Azure subscription ID and resource group name
subscription_id = "<your-subscription-id>"
resource_group_name = "<your-resource-group-name>"

# Set your Azure credentials
credential = DefaultAzureCredential()

# Create the APIM client
apim_client = ApiManagementClient(credential, subscription_id)

# Provision an APIM resource
def provision_apim_resource(resource_name):
    apim_client.api_management.create_or_update(
        resource_group_name,
        resource_name,
        {
            "location": "<your-location>",
            "sku": {
                "name": "Developer",
                "capacity": 1
            }
        }
    )
    print("APIM resource provisioned successfully.")

# List all APIM resources
def list_apim_resources():
    resources = apim_client.api_management.list_by_resource_group(resource_group_name)
    for resource in resources:
        print(resource.name)

# Update an APIM resource
def update_apim_resource(resource_name, new_location):
    apim_client.api_management.update(
        resource_group_name,
        resource_name,
        {
            "location": new_location
        }
    )
    print("APIM resource updated successfully.")

# Delete an APIM resource
def delete_apim_resource(resource_name):
    apim_client.api_management.delete(resource_group_name, resource_name)
    print("APIM resource deleted successfully.")

# Usage examples
provision_apim_resource("<your-apim-resource-name>")
list_apim_resources()
update_apim_resource("<your-apim-resource-name>", "<new-location>")
delete_apim_resource("<your-apim-resource-name>")

File: AZ-204_Implement_API_Management/code/resource_mgmnt.cs
using System;
using Microsoft.Azure.Management.ApiManagement;
using Microsoft.Azure.Management.ApiManagement.Models;
using Microsoft.Rest;

class Program
{
    static void Main(string[] args)
    {
        string subscriptionId = "<Your-Subscription-Id>";
        string resourceGroupName = "<Your-Resource-Group-Name>";
        string serviceName = "<Your-APIM-Service-Name>";

        // Authenticate using your Azure credentials
        var credentials = SdkContext.AzureCredentialsFactory.FromFile("<Your-Azure-Credentials-File>");

        // Create the APIM management client
        var client = new ApiManagementClient(credentials)
        {
            SubscriptionId = subscriptionId
        };

        // Provision an APIM resource
        var apimResource = new ApiManagementServiceResource
        {
            Location = "West US",
            Sku = new ApiManagementServiceSkuProperties
            {
                Name = SkuType.Developer,
                Capacity = 1
            }
        };

        client.ApiManagementService.CreateOrUpdate(resourceGroupName, serviceName, apimResource);

        // List all APIM resources in the subscription
        var apimResources = client.ApiManagementService.ListBySubscription();

        foreach (var resource in apimResources)
        {
            Console.WriteLine($"APIM Resource: {resource.Name}");
        }

        // Update the APIM resource
        var updatedResource = new ApiManagementServiceResource
        {
            Sku = new ApiManagementServiceSkuProperties
            {
                Name = SkuType.Premium,
                Capacity = 2
            }
        };

        client.ApiManagementService.Update(resourceGroupName, serviceName, updatedResource);

        // Delete the APIM resource
        client.ApiManagementService.Delete(resourceGroupName, serviceName);

        Console.WriteLine("APIM resource provisioning, listing, updating, and deleting completed.");
    }
}